crystal_diffusion/score/wrapped_gaussian_score.py

```diff
@@ -0,0 +1,303 @@
+"""Wrapped Gaussian Score.
+
+This module implements the "wrapped Gaussian" score, which corresponds to the "perturbation kernel" on a torus.
+
+The wrapped Gaussian takes the form
+
+    K(x, x0) ~ sum_{k in Z} exp[- |x - x0 + k|^2 / 2 sigma^2], x, x0 in [0, 1).
+
+The score is defined as S = nabla_x ln K(x, x0). Implementing this naively yields an expression that
+potentially converges slowly when sigma is large.
+
+The code below will implement expressions that leverages the "Ewald trick" (ie, part of the sum in real space,
+part of the sum in Fourier space) to insure quick convergence for any sigma. The formula for small
+sigma is derived to avoid division by a number that is very close to zero, or summing very large terms that
+can overlflow.
+
+Also, what is computed is actually "sigma2 x S" (ie, the "sigma normalized score"). This is because
+S ~ 1/ sigma^2; since sigma can be small, this makes the raw score arbitrarily large, and it is better to
+manipulate numbers of small magnitude.
+
+Relevant papers:
+    "Torsional Diffusion for Molecular Conformer Generation",
+        Bowen Jing, Gabriele Corso, Jeffrey Chang, Regina Barzilay, Tommi Jaakkola
+
+    "Riemannian Score-Based Generative Modelling", Bortoli et al.
+"""
+from typing import Union
+
+import numpy as np
+import torch
+
+SIGMA_THRESHOLD = 1.0 / np.sqrt(2.0 * np.pi)
+U_THRESHOLD = 0.5
+
+
+def get_expected_sigma_normalized_score_brute_force(u: float, sigma: float, kmax: Union[float, None] = None):
+    """Brute force implementation.
+
+    A brute force implementation of the sigma normalized score to check that main code is correct.
+    This is only useful if summed to convergence, which is expensive for large sigma!
+    """
+    z = 0.0
+    sigma2_derivative_z = 0.0
+
+    if kmax is None:
+        kmax = np.max([1, np.round(10 * sigma)])
+
+    for k in np.arange(-kmax, kmax + 1):
+        upk = u + k
+        exp = np.exp(-0.5 * upk**2 / sigma**2)
+
+        z += exp
+        sigma2_derivative_z += -upk * exp
+
+    return sigma2_derivative_z / z
+
+
+def get_sigma_normalized_score(
+    relative_positions: torch.Tensor, sigmas: torch.Tensor, kmax: int
+):
+    """Get the sigma normalized score.
+
+    This method branches to different formulas depending on the values of sigma and relative position
+    to insures rapid convergence and numerical stability.
+
+    Args:
+        relative_positions (torch.Tensor): input relative coordinates: should be between 0 and 1.
+            relative_positions is assumed to have an arbitrary shape.
```

Comment:
are we using the convention of repeating the typing in the docstring? We don't need to because they are in the function definition. I'm good with either, just need to decide it

Reply:
I have pycharm set up to use the Google docstring convention
https://google.github.io/styleguide/pyguide.html#doc-function-args

Reply:
I changed the rest of the file based on this.